README.md
[![GitHub](https://img.shields.io/github/license/synthetic-borealis/hersenneuk)](https://github.com/synthetic-borealis/hersenneuk/blob/main/LICENSE)
[![Crates.io](https://img.shields.io/crates/v/hersenneuk)](https://crates.io/crates/hersenneuk)
![Tests](https://github.com/synthetic-borealis/hersenneuk/actions/workflows/ci.yml/badge.svg)

# Hersenneuk

A Brainfuck interpreter written in Rust.

## Usage

### As a standalone program

See ```hersenneuk --help```.

### As a library

Add `hersenneuk` to your crate's dependencies. For other details please see the documentation.


CHANGELOG.md
# Changelog

## v0.2.2
- 20/10/2022 - Added support for C/C++ style multi-line comments.

## v0.2.1

- 13/10/2022 - `run_*`, `get_char` & `put_char` take anything that implements `Read` or `Write` traits.

## v0.2.0

- 12/10/2022 - Added `run_with_dynamic_block`.
- 12/10/2022 - Renamed `run` to `run_with_fixed_block`.
- 12/10/2022 - `run` takes `block_size` as an argument instead of using a constant value.
- 12/10/2022 - `run`, `get_char` & `put_char` take read/write buffer instead of using stdout/stdin directly.


tests/main.rs
use std::io::Write;
use std::process::{Command, Stdio};
use std::{env, path};

#[test]
fn cli_hello_world_fixed_block() {
    let exe_name = env::var("CARGO_PKG_NAME").unwrap();

    #[cfg(target_os = "windows")]
    let exe_name = exe_name + ".exe";

    #[cfg(target_os = "macos")]
    let exe_name = exe_name + ".dmg";

    let target_path =
        env::var("CARGO_MANIFEST_DIR").unwrap() + "/target/debug/" + exe_name.as_str();
    let target_path = path::Path::new(&target_path);
    let output = Command::new(target_path)
        .args(["test_assets/hello_world.bf"])
        .output()
        .unwrap();
    let output = String::from_utf8(output.stdout).unwrap();

    assert_eq!(output.trim(), "Hello World!");
}

#[test]
fn cli_hello_world_dynamic_block() {
    let exe_name = env::var("CARGO_PKG_NAME").unwrap();

    #[cfg(target_os = "windows")]
    let exe_name = exe_name + ".exe";

    #[cfg(target_os = "macos")]
    let exe_name = exe_name + ".dmg";

    let target_path =
        env::var("CARGO_MANIFEST_DIR").unwrap() + "/target/debug/" + exe_name.as_str();
    let target_path = path::Path::new(&target_path);
    let output = Command::new(target_path)
        .args(["test_assets/hello_world.bf", "-D"])
        .output()
        .unwrap();
    let output = String::from_utf8(output.stdout).unwrap();

    assert_eq!(output.trim(), "Hello World!");
}

#[test]
fn cli_user_input_fixed_block() {
    let exe_name = env::var("CARGO_PKG_NAME").unwrap();

    #[cfg(target_os = "windows")]
    let exe_name = exe_name + ".exe";

    #[cfg(target_os = "macos")]
    let exe_name = exe_name + ".dmg";

    let target_path =
        env::var("CARGO_MANIFEST_DIR").unwrap() + "/target/debug/" + exe_name.as_str();
    let target_path = path::Path::new(&target_path);
    let mut child = Command::new(target_path)
        .args(["test_assets/user_input.bf"])
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .spawn()
        .unwrap();

    // ensure child_stdin gets dropped
    {
        let child_stdin = child.stdin.as_mut().unwrap();
        child_stdin.write_all(b"c\n").unwrap();
    }

    let stdout = String::from_utf8(child.wait_with_output().unwrap().stdout).unwrap();
    assert_eq!(stdout.trim(), "c");
}


src/main.rs
use clap::Parser;
use hersenneuk::{code_cleanup, interpreter, syntax_checking};
use std::io::{BufReader, BufWriter};
use std::{fs, io, process};

const BLOCK_SIZE: usize = 30000;
const BUF_SIZE: usize = 1024 * 8;

#[derive(Parser, Debug)]
#[command(author, version, about)]
struct Args {
    /// Input source file
    input_file: String,

    /// Use a dynamic cells block
    #[arg(short = 'D', long, default_value_t = false)]
    with_dynamic_block: bool,
}

fn main() {
    let args = Args::parse();

    if !file_exists(&args.input_file) {
        println!("Error: can't read file {}", args.input_file);
        process::exit(-1);
    }

    let mut stdout = BufWriter::with_capacity(BUF_SIZE, io::stdout());
    let mut stdin = BufReader::with_capacity(BUF_SIZE, io::stdin());
    let mut source_code = fs::read_to_string(args.input_file).expect("Could not read input file!");
    source_code = code_cleanup::clean_code(&code_cleanup::strip_comments(&source_code));

    if syntax_checking::has_mismatching_loop_bounds(&source_code) {
        println!("Error: Mismatching loop bounds detected!");
        process::exit(-1);
    }
    if syntax_checking::has_infinite_loops(&source_code) {
        println!("Error: Potential infinite loop detected!");
        process::exit(-1);
    }

    if args.with_dynamic_block {
        interpreter::run_with_dynamic_block(&source_code, &mut stdin, &mut stdout);
    } else {
        interpreter::run_with_fixed_block(&source_code, &mut stdin, &mut stdout, BLOCK_SIZE);
    }
}

fn file_exists(path: &str) -> bool {
    let metadata = fs::metadata(path);

    match metadata {
        Ok(m) => m.is_file(),
        Err(_e) => false,
    }
}


src/code_cleanup.rs
//! Code cleanup functions.
use regex::Regex;

/// Strips C++-style single-line & multi-line comments.
pub fn strip_comments(code: &str) -> String {
    let multi_line_comment = Regex::new(r"/[*]([^*]|([*][^/]))*[*]/").unwrap();
    let single_line_comment = Regex::new(r"/{2,}[^\r^\n]*").unwrap();
    let first_pass = multi_line_comment.replace_all(code, "").into_owned();
    single_line_comment
        .replace_all(&first_pass, "")
        .into_owned()
}

/// Removes characters that aren't valid Brainfuck instructions.
pub fn clean_code(code: &str) -> String {
    let re = Regex::new(r"[^<>+\-,.\[\]]*").unwrap();
    re.replace_all(code, "").to_string()
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;

    const SOURCE_FILE_WITH_COMMENTS: &str = "test_assets/hello_with_comments.bf";
    const SOURCE_FILE_WITHOUT_COMMENTS: &str = "test_assets/hello_world.bf";

    #[test]
    fn it_cleans_source_correctly() {
        let control_source = fs::read_to_string(SOURCE_FILE_WITHOUT_COMMENTS).unwrap();
        let mut test_source = fs::read_to_string(SOURCE_FILE_WITH_COMMENTS).unwrap();
        test_source = clean_code(&strip_comments(&test_source));
        assert_eq!(control_source, test_source);
    }
}


src/syntax_checking.rs
//! Miscellaneous syntax checking functions.
use regex::Regex;

/// Checks whether a program contains mismatching loop bounds.
pub fn has_mismatching_loop_bounds(code: &str) -> bool {
    let mut loop_count = 0;
    for c in code.chars() {
        if c == '[' {
            loop_count += 1;
        } else if c == ']' {
            loop_count -= 1;
        }
        if loop_count < 0 {
            return true;
        }
    }
    loop_count != 0
}

/// Checks whether a program contains potential infinite loops.
pub fn has_infinite_loops(code: &str) -> bool {
    let re = Regex::new(r"\[[+.]*]").unwrap();
    !matches!(re.find(code), None)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;

    const MISMATCHING_BOUND_SOURCE_FILE: &str = "test_assets/hello_mismatch.bf";
    const INFINITE_LOOP_SOURCE_FILE: &str = "test_assets/inf_loop.bf";

    #[test]
    fn it_detects_mismatching_loop_bounds() {
        let test_source = fs::read_to_string(MISMATCHING_BOUND_SOURCE_FILE).unwrap();
        assert!(has_mismatching_loop_bounds(&test_source));
    }

    #[test]
    fn it_detects_infinite_loops() {
        let test_source = fs::read_to_string(INFINITE_LOOP_SOURCE_FILE).unwrap();
        assert!(has_infinite_loops(&test_source));
    }
}


src/interpreter.rs
//! Brainfuck interpreter functions.
use std::collections::vec_deque::VecDeque;
use std::io::{Read, Write};

/// Runs Brainfuck code with a fixed-size block.
///
/// Note: This function does *NOT* perform any syntax checking.
pub fn run_with_fixed_block(
    code: &str,
    stdin: &mut impl Read,
    stdout: &mut impl Write,
    block_size: usize,
) {
    let instructions: Vec<char> = code.chars().collect();
    let mut cursor: usize = 0;
    let mut cells: Vec<u8> = vec![0; block_size];
    let mut position: usize = 0;
    let mut loop_positions: VecDeque<usize> = VecDeque::new();

    while cursor < instructions.len() {
        match instructions[cursor] {
            '<' => {
                if position > 0 {
                    position -= 1;
                }
            }
            '>' => {
                if position < block_size - 1 {
                    position += 1;
                }
            }
            '-' => {
                if cells[position] > 0 {
                    cells[position] -= 1;
                }
            }
            '+' => {
                if cells[position] < 255 {
                    cells[position] += 1;
                }
            }
            ',' => {
                cells[position] = get_char(stdin) as u8;
            }
            '.' => {
                put_char(cells[position] as char, stdout);
            }
            '[' => {
                if cells[position] != 0 {
                    loop_positions.push_back(cursor);
                } else {
                    cursor += 1;
                    let mut ignore_brace: usize = 1;
                    while cursor < instructions.len() - 1 && ignore_brace != 0 {
                        if instructions[cursor] == ']' {
                            ignore_brace -= 1;
                        } else if instructions[cursor] == '[' {
                            ignore_brace += 1;
                        }
                        cursor += 1;
                    }
                    cursor -= 1;
                }
            }
            ']' => {
                if cells[position] != 0 {
                    cursor = *loop_positions.back().unwrap();
                } else {
                    loop_positions.pop_back();
                }
            }
            _ => {}
        }
        cursor += 1;
    }
}

/// Runs Brainfuck code with a dynamic-sized block.
///
/// Note: This function does *NOT* perform any syntax checking.
pub fn run_with_dynamic_block(code: &str, stdin: &mut impl Read, stdout: &mut impl Write) {
    let instructions: Vec<char> = code.chars().collect();
    let mut cursor: usize = 0;
    let mut cells: Vec<u8> = vec![0];
    let mut position: usize = 0;
    let mut loop_positions: VecDeque<usize> = VecDeque::new();

    while cursor < instructions.len() {
        match instructions[cursor] {
            '<' => {
                if position > 0 {
                    position -= 1;
                }
            }
            '>' => {
                if position + 1 == cells.len() {
                    cells.push(0);
                }
                position += 1;
            }
            '-' => {
                if cells[position] > 0 {
                    cells[position] -= 1;
                }
            }
            '+' => {
                if cells[position] < 255 {
                    cells[position] += 1;
                }
            }
            ',' => {
                cells[position] = get_char(stdin) as u8;
            }
            '.' => {
                put_char(cells[position] as char, stdout);
            }
            '[' => {
                if cells[position] != 0 {
                    loop_positions.push_back(cursor);
                } else {
                    cursor += 1;
                    let mut ignore_brace: usize = 1;
                    while cursor < instructions.len() - 1 && ignore_brace != 0 {
                        if instructions[cursor] == ']' {
                            ignore_brace -= 1;
                        } else if instructions[cursor] == '[' {
                            ignore_brace += 1;
                        }
                        cursor += 1;
                    }
                    cursor -= 1;
                }
            }
            ']' => {
                if cells[position] != 0 {
                    cursor = *loop_positions.back().unwrap();
                } else {
                    loop_positions.pop_back();
                }
            }
            _ => {}
        }
        cursor += 1;
    }
}

fn put_char(c: char, stdout: &mut impl Write) {
    let buf: [u8; 1] = [c as u8];
    stdout.write_all(&buf).unwrap();
}

fn get_char(stdin: &mut impl Read) -> char {
    let mut buf: [u8; 1] = [0];
    stdin.read_exact(&mut buf).unwrap();
    buf[0] as char
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::{BufReader, Cursor};
    use std::{fs, io};

    const HELLO_SOURCE_FILE: &str = "test_assets/hello_world.bf";
    const USER_INPUT_SOURCE_FILE: &str = "test_assets/user_input.bf";
    const BLOCK_SIZE: usize = 30000;
    const BUF_SIZE: usize = 1024 * 8;

    struct MockInput {
        buffer: Vec<u8>,
    }

    impl MockInput {
        pub fn from_str(s: &str) -> Self {
            Self {
                buffer: Vec::from(s),
            }
        }
    }

    impl Read for MockInput {
        fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
            let mut r = Cursor::new(&self.buffer);
            r.read(buf)
        }
    }

    #[test]
    fn hello_world_fixed_block() {
        let test_source = fs::read_to_string(HELLO_SOURCE_FILE).unwrap();
        let mut stdout: Vec<u8> = Vec::new();
        let mut stdin = BufReader::with_capacity(BUF_SIZE, io::stdin());

        run_with_fixed_block(&test_source, &mut stdin, &mut stdout, BLOCK_SIZE);
        let stdout = String::from_utf8_lossy(stdout.as_slice()).to_string();

        assert_eq!(stdout.trim(), "Hello World!");
    }

    #[test]
    fn hello_world_dynamic_block() {
        let test_source = fs::read_to_string(HELLO_SOURCE_FILE).unwrap();
        let mut stdout: Vec<u8> = Vec::new();
        let mut stdin = BufReader::with_capacity(BUF_SIZE, io::stdin());

        run_with_dynamic_block(&test_source, &mut stdin, &mut stdout);
        let stdout = String::from_utf8_lossy(stdout.as_slice()).to_string();

        assert_eq!(stdout.trim(), "Hello World!");
    }

    #[test]
    fn user_input() {
        let test_source = fs::read_to_string(USER_INPUT_SOURCE_FILE).unwrap();
        let mut stdout: Vec<u8> = Vec::new();
        let mut stdin = MockInput::from_str("c\n");

        run_with_fixed_block(&test_source, &mut stdin, &mut stdout, BLOCK_SIZE);
        let stdout = String::from_utf8_lossy(stdout.as_slice()).to_string();

        assert_eq!(stdout.trim(), "c");
    }
}
